src/core/execute.c

```diff
@@ -720,74 +720,129 @@ static int ask_for_confirmation(char *response, char **argv) {
+                                          const char *user,
+                                          const char *group,
+                                          gid_t gid, gid_t **supplementary_gids) {
+        char **i;
+        int r, k = 0;
+        int ngroups_max;
         bool keep_groups = false;
-        int r;
+        gid_t *l_gids;
 
-        assert(context);
+        assert(c);
 
-        /* Lookup and set GID and supplementary group list. Here too
-         * we avoid NSS lookups for gid=0. */
+        if (!c->supplementary_groups)
+                return 0;
 
-        if (context->group || username) {
+        /* Lookup GID and supplementary group list. We avoid NSS lookups for
+         * gid=0. */
+        if (user && group && gid != GID_INVALID && gid != 0) {
                 /* First step, initialize groups from /etc/groups */
-                if (username && gid != 0) {
-                        if (initgroups(username, gid) < 0)
-                                return -errno;
+                if (initgroups(user, gid) < 0)
+                        return -errno;
 
-                        keep_groups = true;
-                }
+                keep_groups = true;
+        }
 
-                /* Second step, set our gids */
-                if (setresgid(gid, gid, gid) < 0)
+        assert_se((ngroups_max = (int) sysconf(_SC_NGROUPS_MAX)) > 0);
+
+        l_gids = new(gid_t, ngroups_max);
+        if (!l_gids)
+                return -ENOMEM;
+
+        if (keep_groups) {
+                k = getgroups(ngroups_max, l_gids);
+                if (k < 0) {
+                        free(l_gids);
                         return -errno;
+                }
+        } else
+                k = 0;
+
+        STRV_FOREACH(i, c->supplementary_groups) {
+                const char *g;
+
+                if (k >= ngroups_max) {
+                        free(l_gids);
+                        return -E2BIG;
+                }
+
+                g = *i;
+                r = get_group_creds(&g, l_gids+k);
+                if (r < 0) {
+                        free(l_gids);
+                        return r;
+                }
+
+                k++;
         }
 
-        if (context->supplementary_groups) {
-                int ngroups_max, k;
-                gid_t *gids;
-                char **i;
+        if (k >= ngroups_max) {
+                free(l_gids);
+                return -E2BIG;
+        }
 
-                /* Final step, initialize any manually set supplementary groups */
-                assert_se((ngroups_max = (int) sysconf(_SC_NGROUPS_MAX)) > 0);
+        /* Close gids and GID_INVALID can't be used from user passed values
+         * in any case we will fail before if we can't resolve a group */
+        l_gids[k] = GID_INVALID;
 
-                if (!(gids = new(gid_t, ngroups_max)))
-                        return -ENOMEM;
+        *supplementary_gids = l_gids;
```

Comment:
I'd probably declare l_gids with `_cleanup_free_` so that we can drop the explicit free() calls in the error clean-up paths. Of course, after assigning it to *supplementary_gids we'd then have to reset it to NULL, to make sure _cleanup_free_ doesn't break.